Added BIND_AS_AUTHENTICATING_USER option.

vitormazzi opened the pull request:
This option allows the authentication of users with multiple unknown
distinguished name (dn) formats, provided that there is at least one
user whose dn is known and has the permissions to search for other
users.

vitormazzi committed:
Added BIND_AS_AUTHENTICATING_USER option.

This option allows the authentication of users with multiple unknown
distinguished name (dn) formats, provided that there is at least one
user whose dn is known and has the permissions to search for other
users.

vitormazzi committed:
bugfix: Handling bind failures correctly

sjkingo commented:
Thanks for your PR. Could you explain a little how this is different to method 2 (search and bind), except with a custom filter?

sjkingo commented:
It would be awesome to merge this into the code as the search and bind method (issue #2) as that has been a long-standing feature I've wanted to implement.

vitormazzi commented:
I don't have a lot of ldap knowledge, so I'm not really sure about the differences. That is the main reason I proposed this code as a third authentication method.

The only incompatible point I see at this moment is that the search step in search and bind could be anonymous (is that right?), and in my code it requires a valid password. Do you see anything else?

vitormazzi committed:
Updating for `ldap3 == 2.1.1`

Replacing references to `EntryBase.entry_get_dn()` with
`EntryBase.entry_dn`, as this method does not exist anymore.

Using a specific version of ldap3 from now on, as this api change
caught us off guard.
